File: android/src/main/kotlin/com/meludia/FlutterPDTest/FlutterPd.kt
package com.meludia.FlutterPDTest

import android.content.Context
import com.meludia.FlutterPDTest.exception.PdException
import com.meludia.FlutterPDTest.util.RequestPermissionHandler
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import java.io.File
import java.io.IOException

class FlutterPd(private val context: Context,
                private val assetPathResolver: AssetPathResolver,
                private val pd: Pd) : DartToNative {
  var activityBinding: ActivityPluginBinding? = null

  fun dispose() {
    pd.release()
  }

  override fun onListen(symbol: String, id: Int, callback: (Any) -> Unit) {
    pd.addListener(symbol, PdEventListener(id, callback))
  }

  override fun onCancel(symbol: String, id: Int) {
    PdEventListener.remove(id)?.let { listener ->
      pd.removeListener(symbol, listener)
    }
  }

  override fun checkPermission(callback: (Boolean) -> Unit) {
    val activityBinding = this.activityBinding
        ?: throw PdException("Plugin is not bound to Flutter", "requestPermission failed")
    RequestPermissionHandler(callback).let {
      if (it.requestPermission(activityBinding.activity)) {
        activityBinding.addRequestPermissionsResultListener(it)
      }
    }
  }

  override fun startPd(callback: () -> Unit) {
    if (pd.isRunning || pd.isPreparing) {
      throw  PdException("Already started", "start is called more than twice")
    }
    pd.startService(callback)
  }

  override fun stopPd() {
    if (pd.isRunning) {
      pd.stopService()
    }
  }

  override fun openPatch(patchFilePath: String): Int {
    try {
      val patchFile = File(context.cacheDir, patchFilePath)
      return pd.openPatch(patchFile)
    } catch (e: IOException) {
      throw PdException("openPatch failed", e.message)
    }
  }

  override fun close(patchHandle: Int) {
    pd.closePatch(patchHandle)
  }

  override fun startAudio(requireInput: Boolean) {
    if (pd.isRunning.not()) {
      throw PdException("PdService is not started", "call start() first.")
    }

    return try {
      pd.initAudio(requireInput)
      pd.startAudio()
    } catch (e: Exception) {
      throw PdException("startAudio failed", e.message, null)
    }
  }

  override fun send(receiver: String, value: Any) {
    if (pd.isRunning.not()) {
      throw PdException("PdService is not started", "call start() first.")
    }

    pd.send(receiver, value)
  }
}

File: lib/flutter_pd.dart
import 'flutter_pd_platform_interface.dart';
import 'pd_event.dart';
import 'pd_file_handle.dart';

export 'pd_event.dart';
export 'pd_file_handle.dart';

class FlutterPd {
  FlutterPd._();
  static final FlutterPd instance = FlutterPd._();
  /// Request a permission to access to microphone.
  Future<bool> checkPermission() async {
    return FlutterPdPlatform.instance.checkPermission();
  }

  Future<void> startPd() {
    return FlutterPdPlatform.instance.startPd();
  }

  Future<void> restartAudio(bool requestInput){
    return FlutterPdPlatform.instance.restartAudio(requestInput);
  }

  Future<void> stopPd() {
    return FlutterPdPlatform.instance.stopPd();
  }

  Future<PdFileHandle> openPatch(String patchFilePath) async {
    return FlutterPdPlatform.instance.openPatch(patchFilePath);
  }

  /// Should only be called from [PdFileHandle.close]. Do not call this directly.
  Future<void> close(int patchHandle) {
    return FlutterPdPlatform.instance.close(patchHandle);
  }

  Future<void> startAudio({
    bool? requireInput,
  }) {
    return FlutterPdPlatform.instance.startAudio(requireInput: requireInput);
  }

  Future<void> send(String receiverName, dynamic value) {
    return FlutterPdPlatform.instance.send(receiverName, value);
  }

  Stream<PdEvent> receive(String symbol) {
    return FlutterPdPlatform.instance.receive(symbol);
  }
}
